dbus-wrapper/swupdd-main.c

```diff
@@ -0,0 +1,985 @@
+		args = list_append_data(args, strdup(bundle));
+	}
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Can't read bundle name");
+		goto finish;
+	}
+	r = sd_bus_message_exit_container(m);
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Can't exit bundles container");
+		goto finish;
+	}
+
+	r  = run_swupd(METHOD_BUNDLE_ADD, args, context);
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Failed to run swupd command");
+		goto finish;
+	}
+
+	r = sd_bus_reply_method_return(m, "b", (r >= 0));
+
+finish:
+	list_free_list_and_data(args, free);
+	return r;
+}
+
+static int method_bundle_remove(sd_bus_message *m,
+				void *userdata,
+				sd_bus_error *ret_error)
+{
+	daemon_state_t *context = userdata;
+	int r = 0;
+	struct list *args = NULL;
+
+	r = check_prerequisites(context, ret_error);
+	if (r < 0) {
+		return r;
+	}
+
+	args = list_append_data(args, strdup(SWUPD_CLIENT));
+	args = list_append_data(args, strdup(_method_opt_map[METHOD_BUNDLE_REMOVE]));
+
+	char const * const str_opts[] = {"path", "url", "contenturl", "versionurl",
+					 "format", "statedir", NULL};
+	char const * const bool_opts[] = {"force", NULL};
+	char const * const int_opts[] = {"port", NULL};
+	r = bus_message_read_options(m, str_opts, bool_opts, int_opts, &args, ret_error);
+	if (r < 0) {
+		goto finish;
+	}
+
+	const char *bundle;
+	r = sd_bus_message_read(m, "s", &bundle);
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Can't read bundle name");
+		goto finish;
+	}
+	args = list_append_data(args, strdup(bundle));
+
+	r  = run_swupd(METHOD_BUNDLE_REMOVE, args, context);
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Failed to run swupd command");
+		goto finish;
+	}
+
+	r = sd_bus_reply_method_return(m, "b", (r >= 0));
+
+finish:
+	list_free_list_and_data(args, free);
+	return r;
+}
+
+static int method_cancel(sd_bus_message *m,
+			 void *userdata,
+			 sd_bus_error *ret_error)
+{
+	daemon_state_t *context = userdata;
+	int r = 0;
+	pid_t child;
+	int force;
+
+	child = context->child;
+	if (!child) {
+		sd_bus_error_set_errnof(ret_error, ECHILD, "No child process to cancel");
+		return -ECHILD;
+	}
+
+	r = sd_bus_message_read(m, "b", &force);
+	if (r < 0) {
+		sd_bus_error_set_errnof(ret_error, -r, "Can't read 'force' option");
+		return r;
+	}
+
+	if (force) {
+		kill(child, SIGKILL);
+	} else {
+		kill(child, SIGTERM);
+	}
+
+	return sd_bus_reply_method_return(m, "b", (r >= 0));
+}
+
+/* Basically this is a copypasta from systemd's internal function bus_event_loop_with_idle() */
+static int run_bus_event_loop(sd_event *event,
+			      daemon_state_t *context)
+{
+	sd_bus *bus = context->bus;
+	bool exiting = false;
+	int r, code;
+
+	for (;;) {
+		r = sd_event_get_state(event);
+		if (r < 0) {
+			ERR("Failed to get event loop's state: %s", strerror(-r));
+			return r;
+		}
+		if (r == SD_EVENT_FINISHED) {
+			break;
+		}
+
+		r = sd_event_run(event, exiting ? (uint64_t) -1 : (uint64_t) TIMEOUT_EXIT_SEC * 1000000);
```

Comment:
I'm not seeing it documented well but what does sd_event_run return when it hits the timeout?